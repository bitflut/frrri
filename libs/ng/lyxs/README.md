# ng-lyxs

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test ng-lyxs` to execute the unit tests.
