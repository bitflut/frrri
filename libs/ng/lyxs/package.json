{
  "$schema": "../../../node_modules/ng-packagr/package.schema.json",
  "name": "@lyxs/angular",
  "version": "0.0.1",
  "license": "MIT",
  "private": true,
  "peerDependencies": {
    "@angular/common": "^9.1.0",
    "@angular/core": "^9.1.0",
    "@ngxs/store": "^3.6.2",
    "@ngxs-labs/data": "^3.0.4"
  },
  "keywords": [
    "Angular",
    "LYXS",
    "NGXS",
    "API",
    "CRUD",
    "State management"
  ],
  "author": "marian@bitflut.com bitflut.com",
  "homepage": "https://bitflut.gitbook.io/lyxs/",
  "repository": {
    "type": "git",
    "url": "https://github.com/bitflut/lyxs"
  },
  "ngPackage": {
    "lib": {
      "entryFile": "src/public_api.ts"
    },
    "dest": "../../../dist/@lyxs/angular"
  }
}
