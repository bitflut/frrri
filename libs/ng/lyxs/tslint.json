{
  "extends": "../../../tslint.json",
  "rules": {
    "directive-selector": [true, "attribute", "lyxs", "camelCase"],
    "component-selector": [true, "element", "lyxs", "kebab-case"]
  },
  "linterOptions": {
    "exclude": ["!**/*"]
  }
}
