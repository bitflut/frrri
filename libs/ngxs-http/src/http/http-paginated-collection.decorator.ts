import { CollectionReducer } from '@frrri/ngxs';
import { HttpCollection } from './http-collection.decorator';
import { StateClass } from '@ngxs/store/internals';
import { PaginatedCollectionOptions } from '@frrri/ngxs/pagination';
import { HttpPaginatedCollectionService } from './http-paginated-collection.service';

export function HttpPaginatedCollection<T = CollectionReducer>(options: PaginatedCollectionOptions<T>) {
    options = {
        ...options,
        defaults: {
            next: undefined,
            ...options.defaults,
        },
    };

    const crudCollectionFn = HttpCollection(options);
    return function (target: StateClass) {
        target.prototype.pageSize = options.size;
        target.prototype.paginatedServiceToken = HttpPaginatedCollectionService;
        crudCollectionFn(target);
    };
}
