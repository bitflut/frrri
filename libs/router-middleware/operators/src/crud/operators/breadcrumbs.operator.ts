import { Platform } from '@frrri/router-middleware';
import { Operation } from '../../interfaces/operation.interface';
import { CrudOperatorType } from '../enums/crud-operator-type';

let index = -1;

interface StaticOptions {
    title: string;
    id: number;
}

interface ActiveOptions<T> {
    id: number;
    factory: (data: T) => Omit<StaticOptions, 'id'>;
}

export function staticBreadcrumb(
    statePath: string,
    options: StaticOptions,
) {
    index++;
    return {
        type: CrudOperatorType.StaticBreadcrumb as CrudOperatorType.StaticBreadcrumb,
        options: { ...options, id: index },
        statePath,
        platforms: [Platform.NavigationEnd],
    } as Operation;
}

export function activeBreadcrumb<T = any>(
    statePath: string,
    options: ActiveOptions<T>,
) {
    index++;
    return {
        type: CrudOperatorType.ActiveBreadcrumb as CrudOperatorType.ActiveBreadcrumb,
        options: { ...options, id: index },
        statePath,
        platforms: [Platform.NavigationEnd],
    } as Operation;
}
