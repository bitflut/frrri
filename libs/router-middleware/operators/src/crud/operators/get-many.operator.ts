import { Platform } from '@frrri/router-middleware';
import { Operation } from '../../interfaces/operation.interface';
import { crudOperatorDefaults } from '../defaults/crud-operator.defaults';
import { CrudOperatorType } from '../enums/crud-operator-type';

export function getMany(
    statePath: string,
    options: {
        /** Params passed to getMany */
        params?: { [key: string]: string | string[] };
    } = {},
) {
    options = {
        ...crudOperatorDefaults,
        params: {},
        ...options,
    };

    return {
        type: CrudOperatorType.GetMany as CrudOperatorType.GetMany,
        statePath,
        platforms: [Platform.Resolver],
        ...options,
    } as Operation;
}
