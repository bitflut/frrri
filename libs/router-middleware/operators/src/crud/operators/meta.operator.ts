import { Platform } from '@frrri/router-middleware';
import { Operation } from '../../interfaces/operation.interface';
import { CrudOperatorType } from '../enums/crud-operator-type';

export interface MetaOptions {
    title?: string;
    keywords?: string;
    description?: string;
    image?: string;
}

export interface ActiveMetaOptions<T> {
    factory: (data: T) => MetaOptions;
}

export function activeMeta<T = any>(
    statePath: string,
    options: ActiveMetaOptions<T>,
) {
    return {
        type: CrudOperatorType.ActiveMeta as CrudOperatorType.ActiveMeta,
        options,
        statePath,
        platforms: [Platform.NavigationEnd],
    } as Operation;
}

export function staticMeta(
    statePath: string,
    options: MetaOptions = {},
) {
    return {
        type: CrudOperatorType.StaticMeta as CrudOperatorType.StaticMeta,
        options,
        statePath,
        platforms: [Platform.NavigationEnd],
    } as Operation;
}
