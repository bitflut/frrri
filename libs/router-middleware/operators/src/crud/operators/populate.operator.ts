import { Platform } from '@frrri/routes';
import { Operation } from '../../interfaces/operation.interface';
import { CrudOperatorType } from '../enums/crud-operator-type';

export function populate(
    options: {
        /** State path that triggers a population */
        from: string,
        /** State path getting populated */
        to: string,
        /** Path of id property */
        idPath?: string,
        /** State path of collection containing id (foreign vs self) */
        idSource?: string,
        /**
         * Used to map ids to getMany params when populating.
         * Default:
         * ```
         * (ids: IdType[], path: string) => {
         *     return {
         *          [`filter[${path}][$in]`]: uniq(ids),
         *     }
         * }
         * ```
         */
        populatFactory?: (ids: string | number[], path: string) => any;
    },
) {
    return {
        type: CrudOperatorType.Populate as CrudOperatorType.Populate,
        statePath: options.from,
        toStatePath: options.to,
        idPath: options.idPath ?? 'id',
        idSource: options.idSource ?? options.from,
        platforms: [Platform.Resolver],
        ...options,
    } as Operation;
}
