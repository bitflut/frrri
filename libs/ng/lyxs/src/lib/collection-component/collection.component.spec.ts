import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { Injectable } from '@angular/core';
import { ComponentFixture, inject, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { Cache, CacheState, CrudCollection, CrudCollectionState, PaginatedCrudCollectionState, PaginationInterceptor } from '@lya/ng/lyxs';
import { NgxsDataPluginModule } from '@ngxs-labs/data';
import { NgxsModule } from '@ngxs/store';
import { MockRender } from 'ng-mocks';
import { take } from 'rxjs/operators';
import { CollectionComponent } from './collection.component';
import { LyxsCollectionModule } from './collection.module';

interface Post {
    userId: number;
    id: number;
    body: string;
    title: string;
}

@CrudCollection({
    baseUrl: 'https://jsonplaceholder.typicode.com',
    name: 'posts',
})
@Injectable()
class PostsEntitiesState extends PaginatedCrudCollectionState<Post, number> { }

interface Comment {
    postId: number;
    id: number;
    name: string;
    body: string;
    email: string;
}

@CrudCollection({
    name: 'comments',
    baseUrl: 'https://jsonplaceholder.typicode.com',
})
@Injectable()
class CommentsEntitiesState extends CrudCollectionState<Comment, number> { }

@Cache({
    name: 'cache',
    defaults: {},
    children: [PostsEntitiesState, CommentsEntitiesState],
})
@Injectable()
class EntityCacheState extends CacheState<any> { }

const page1Data = {
    body: [{
        userId: 1,
        id: 1,
        body: 'Hello World',
        title: 'testing Angular',
    }, {
        userId: 2,
        id: 2,
        body: 'Hello World2',
        title: 'testing Angular2',
    }],
    headers: {
        Link: '</api/v1/admin/posts/?sort=name&pageSize=20&next%5B%24or%5D%5B0%5D%5Bname%5D%5B%24gt%5D=Forum%20Mannheim&next%5B%24or%5D%5B1%5D%5Bname%5D%5B%24gte%5D=Forum%20Mannheim&next%5B%24or%5D%5B1%5D%5B_id%5D%5B%24gt%5D=5e4d2901871f09a8d0fc4bef>; rel="next"',
    },
};

const page2Data = {
    body: [{
        userId: 1,
        id: 3,
        body: 'Hello World3',
        title: 'testing Angular3',
    }, {
        userId: 2,
        id: 4,
        body: 'Hello World4',
        title: 'testing Angular4',
    }],
    headers: {
        Link: 'link: </api/v1/admin/stores/?sort=name&pageSize=20&next%5B%24or%5D%5B0%5D%5Bname%5D%5B%24gt%5D=LARI%20LUKE&next%5B%24or%5D%5B1%5D%5Bname%5D%5B%24gte%5D=LARI%20LUKE&next%5B%24or%5D%5B1%5D%5B_id%5D%5B%24gt%5D=5cffc130c82fe51c5afe0a9e>; rel="next"',
    },
};

const page3Data = {
    body: [{
        userId: 1,
        id: 5,
        body: 'Hello World3',
        title: 'testing Angular3',
    }],
    headers: {},
};

describe('CollectionComponent', () => {
    let component: CollectionComponent;
    let fixture: ComponentFixture<CollectionComponent>;

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [
                HttpClientTestingModule,
                NgxsModule.forRoot([EntityCacheState, PostsEntitiesState, CommentsEntitiesState]),
                NgxsDataPluginModule.forRoot(),
                LyxsCollectionModule,
            ],
            providers: [{
                provide: HTTP_INTERCEPTORS,
                multi: true,
                useClass: PaginationInterceptor,
            }],
        }).compileComponents();

        fixture = MockRender(`
            <lyxs-collection path="cache.posts">
                My content
                <div class="loading">mock-loading</div>
                <div class="loading-first">mock-first-load</div>
                <div class="load-more">mock-more</div>
            </lyxs-collection>
        `);

        component = fixture.debugElement
            .query(By.directive(CollectionComponent))
            .componentInstance as CollectionComponent;
    });

    it('should create and initialize', async () => {
        expect(component).toBeTruthy();
        const loading = await component.loading$.pipe(take(1)).toPromise();
        expect(loading).toBeFalsy();
    });

    it('should show contents correctly', inject([
        HttpTestingController,
        PostsEntitiesState,
    ], (
        httpMock: HttpTestingController,
        postsEntities: PostsEntitiesState,
    ) => {
        // INIT
        expect(fixture.nativeElement.textContent.trim()).toEqual('My content');
        expect(fixture).toMatchSnapshot('init');

        // PAGE 1/3
        postsEntities.getMany().toPromise();
        const req1 = httpMock.expectOne(postsEntities.requestOptions.collectionUrlFactory());
        fixture.detectChanges();

        // PAGE 1/3 - LOADING FIRST
        expect(fixture).toMatchSnapshot('loading-page-1-of-3');
        req1.flush(page1Data.body, { headers: page1Data.headers });
        fixture.detectChanges();

        // PAGE 1/3 - LOADED
        expect(fixture).toMatchSnapshot('loaded-page-1-of-3');

        // PAGE 2/3
        postsEntities.getNext().toPromise();
        const req2 = httpMock.expectOne(`${postsEntities.requestOptions.collectionUrlFactory()}?sort=name&pageSize=20&next%5B$or%5D%5B0%5D%5Bname%5D%5B$gt%5D=Forum%20Mannheim&next%5B$or%5D%5B1%5D%5Bname%5D%5B$gte%5D=Forum%20Mannheim&next%5B$or%5D%5B1%5D%5B_id%5D%5B$gt%5D=5e4d2901871f09a8d0fc4bef`);
        fixture.detectChanges();

        // PAGE 2/3 - LOADING
        expect(fixture).toMatchSnapshot('loading-page-2-of-3');
        req2.flush(page2Data.body, { headers: page2Data.headers });
        fixture.detectChanges();

        // PAGE 2/3 LOADED
        expect(fixture).toMatchSnapshot('loaded-page-2-of-3');

        // PAGE 3/3 (LAST PAGE)
        postsEntities.getNext().toPromise();
        const req3 = httpMock.expectOne(`${postsEntities.requestOptions.collectionUrlFactory()}?sort=name&pageSize=20&next%5B$or%5D%5B0%5D%5Bname%5D%5B$gt%5D=LARI%20LUKE&next%5B$or%5D%5B1%5D%5Bname%5D%5B$gte%5D=LARI%20LUKE&next%5B$or%5D%5B1%5D%5B_id%5D%5B$gt%5D=5cffc130c82fe51c5afe0a9e`);
        fixture.detectChanges();

        // PAGE 3/3 LOADING
        expect(fixture).toMatchSnapshot('loading-page-3-of-3');
        req3.flush(page3Data.body, { headers: page3Data.headers });
        fixture.detectChanges();

        // PAGE 3/3 LOADED
        expect(fixture).toMatchSnapshot('loaded-page-3-of-3');
    }));

    it('should show errors', inject([
        HttpTestingController,
        PostsEntitiesState,
    ], (
        httpMock: HttpTestingController,
        postsEntities: PostsEntitiesState,
    ) => {
        // INIT
        expect(fixture.nativeElement.textContent.trim()).toEqual('My content');
        expect(fixture).toMatchSnapshot('init');

        // PAGE 1/3
        postsEntities.getMany().toPromise();
        const req1 = httpMock.expectOne(postsEntities.requestOptions.collectionUrlFactory());
        fixture.detectChanges();

        // PAGE 1/3 - LOADING FIRST
        expect(fixture).toMatchSnapshot('loading-page-1-of-3');
        req1.flush({}, { status: 400, statusText: 'Bad Request' });
        fixture.detectChanges();

        // PAGE 1/3 - ERROR
        expect(fixture).toMatchSnapshot('error-page-1-of-3');
    }));

    afterEach(inject([HttpTestingController], (httpMock: HttpTestingController) => {
        httpMock.verify();
    }));
});
