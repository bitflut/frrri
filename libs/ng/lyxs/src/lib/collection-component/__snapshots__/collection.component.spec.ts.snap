// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionComponent should show contents correctly: init 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show contents correctly: loaded-page-1-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
    <div
      class="load-more-content"
    >
      <div>
        <div>
          <div
            class="load-more"
          >
            mock-more
          </div>
        </div>
        
      </div>
    </div>
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show contents correctly: loaded-page-2-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
    <div
      class="load-more-content"
    >
      <div>
        <div>
          <div
            class="load-more"
          >
            mock-more
          </div>
        </div>
        
      </div>
    </div>
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show contents correctly: loaded-page-3-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show contents correctly: loading-page-1-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    <div
      class="loading-first-content"
    >
      <div>
        <div
          class="loading-first"
        >
          mock-first-load
        </div>
      </div>
      
    </div>
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show contents correctly: loading-page-2-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    <div
      class="loading-content"
    >
      <div>
        <div
          class="loading"
        >
          mock-loading
        </div>
      </div>
    </div>
    
    
    
    
    
    <div
      class="loading-more-content"
    >
      <div />
      
      <div>
         Loading more... 
      </div>
    </div>
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show contents correctly: loading-page-3-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    <div
      class="loading-content"
    >
      <div>
        <div
          class="loading"
        >
          mock-loading
        </div>
      </div>
    </div>
    
    
    
    
    
    <div
      class="loading-more-content"
    >
      <div />
      
      <div>
         Loading more... 
      </div>
    </div>
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show errors: error-page-1-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    <div
      class="error-content"
    >
      <div />
      
      
      <div
        style="color: red;"
      >
        Http failure response for https://jsonplaceholder.typicode.com/posts: 400 Bad Request
      </div>
    </div>
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show errors: init 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;

exports[`CollectionComponent should show errors: loading-page-1-of-3 1`] = `
<mock-render>
  <lyxs-collection
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    <div
      class="loading-first-content"
    >
      <div>
        <div
          class="loading-first"
        >
          mock-first-load
        </div>
      </div>
      
    </div>
    
    
    
    
    
  </lyxs-collection>
</mock-render>
`;
